Add bucket types

Vagabond opened the pull request:
see also, basho/riak#362

Vagabond committed:
WIP bucket types

Vagabond committed:
Make riak_object speak bucket types

Vagabond committed:
Sanitize the default bucket type out at the vnode layer

The vnode is unlikely to see this unless the user is using
riak:local_client, but better to be safe.

Vagabond committed:
More API changes to support bucket types

Vagabond committed:
Teach mapreduce about bucket types

Vagabond committed:
Fix bucket folding for bucket-types and eleveldb

engelsanchez commented on src/riak_client.erl:
```diff
@@ -375,8 +375,11 @@ stream_list_keys(Bucket, Timeout, {?MODULE, [_Node, _ClientId]}=THIS) ->
 %%      keys in Bucket on any single vnode.
 stream_list_keys(Input, Timeout, Client, {?MODULE, [Node, _ClientId]}) when is_pid(Client) ->
     ReqId = mk_reqid(),
+    lager:info("input is ~p", [Input]),
```
Don't forget to remove this!

engelsanchez commented:
I don't think I've seen any typos, but consider using a macro instead of peppering <<"default">> around just in case.

engelsanchez commented on src/riak_kv_pb_bucket.erl:
```diff
@@ -91,9 +91,10 @@ process(rpblistbucketsreq, State) ->
+process(#rpblistkeysreq{type = Type, bucket=B,timeout=T}=Req, #state{client=C} = State) ->
     %% stream_list_keys results will be processed by process_stream/3
-    {ok, ReqId} = C:stream_list_keys(B, T),
+    Bucket = maybe_create_bucket_type(Type, B),
```
Not a real issue, but I got confused for a second when I read this the first time thinking this could try to create a bucket type. Consider a name change maybe.

engelsanchez commented on src/riak_object.erl:
```diff
@@ -263,6 +273,17 @@ apply_updates(Object=#r_object{}) ->
+%% @doc Return the containing bucket-type for this riak_object.
+-spec type(riak_object()) -> binary() | undefined.
+type(#r_object{bucket={Type, _Bucket}}) -> Type;
+type(#r_object{bucket=_Bucket}) -> undefined.
```
How about using bucket_type instead of type here? It could get confusing. It sounds like object type.

engelsanchez commented on src/riak_object_json.erl:
```diff
@@ -33,7 +33,8 @@
 encode(Obj) ->
     {_,Vclock} = riak_object:vclock_header(Obj),
-    {struct, [{<<"bucket">>, riak_object:bucket(Obj)},
+    {struct, [{<<"type">>, riak_object:type(Obj)},
```
Same here. I think it would be less confusing to use bucket_type (or btype or w/e) instead of plain type at the object level.

engelsanchez commented on src/riak_kv_pipe_index.erl:
```diff
@@ -105,7 +105,7 @@ keysend(_Bucket, [], _Partition, _FittingDetails) ->
 keysend(Bucket, [Key | Keys], Partition, FittingDetails) ->
     case riak_pipe_vnode_worker:send_output(
-           {Bucket, strip_index(Key)}, Partition, FittingDetails) of
+           {{Bucket, strip_index(Key)}, undefined}, Partition, FittingDetails) of
```
This makes the mapred_basic_compat riak test fail. Do we need to change this?

engelsanchez commented:
@Vagabond please see my note above about the MR riak test failing because it's expecting {Bucket, Key} and it's now getting {{Bucket, Key}, undefined}. I'm not following the need for that change. Could you explain?

Vagabond committed:
Fix bug where queue_existing_pipe did not decode bucket correctly

When a key-filter was being used, pipe would do a get_bucket with the
{Bucket, Filter} tuple, which would return the default bucket
properties, thus potentially using the wrong n_val for the coverage
query.

Vagabond committed:
Fix bucket folding when bucket type is undefined

Vagabond committed:
Fix guard to avoid passing undefined as Client pid
